### ISSUE TITLE
[BUG] iscsid - connect to x.x.x.x failed - no route to host

### ISSUE THREAD

**Issue by dominch** [longhorn/longhorn#7386]
## Describe the bug (🐛 if you encounter this issue)

On some nodes I have several iscsi connections broken, on iscsid service there are logged such messages: 

```
Dec 19 13:19:36 k3s-node-2 iscsid[3160778]: connect to 10.42.3.235:3260 failed (No route to host)
Dec 19 13:19:44 k3s-node-2 iscsid[3160778]: connect to 10.42.3.235:3260 failed (No route to host)
Dec 19 13:19:51 k3s-node-2 iscsid[3160778]: connect to 10.42.3.235:3260 failed (No route to host)
Dec 19 13:19:58 k3s-node-2 iscsid[3160778]: connect to 10.42.3.235:3260 failed (No route to host)
Dec 19 13:20:05 k3s-node-2 iscsid[3160778]: connect to 10.42.3.235:3260 failed (No route to host)
Dec 19 13:20:12 k3s-node-2 iscsid[3160778]: connect to 10.42.3.235:3260 failed (No route to host)
```

As in timestamp - this is logged about every 5-15s 
There is no such ip in k3s, on other node there are two wrong endpoints. Message comes from iscsid service but somehow is connected with longhorn. 
I tried to remove those from discovered hosts on iscsid service but restarting it does not help, when I reboot node they usually gone. 
I also see this ip in instance manager logs: 

```
containers/instance-manager-554c586ceba126a4394b9acba2421017_longhorn-system_instance-manager-d21cda831661a540310183bab441cdd0b7fa8504cbc
3e1bc1341186afd9ea2d9.log:2023-11-06T14:00:50.233682662+01:00 stderr F [pvc-051baae1-4be6-4fa0-b21a-859bf04f5a67-r-03698787] time="2023-1
1-06T13:00:50Z" level=info msg="New connection from: 10.42.3.235:46206" func="rpc.(*DataServer).listenAndServeTCP" file="dataserver.go:57
"
containers/instance-manager-554c586ceba126a4394b9acba2421017_longhorn-system_instance-manager-d21cda831661a540310183bab441cdd0b7fa8504cbc
3e1bc1341186afd9ea2d9.log:2023-11-06T14:00:50.240321222+01:00 stderr F [pvc-2e73dd4b-4afd-428c-9ae0-a4032eab8246-r-c0f6fda1] time="2023-1
1-06T13:00:50Z" level=info msg="New connection from: 10.42.3.235:51032" func="rpc.(*DataServer).listenAndServeTCP" file="dataserver.go:57
"
containers/instance-manager-554c586ceba126a4394b9acba2421017_longhorn-system_instance-manager-d21cda831661a540310183bab441cdd0b7fa8504cbc
3e1bc1341186afd9ea2d9.log:2023-11-06T14:00:50.894585699+01:00 stderr F [pvc-5cd535db-14b8-41a6-b689-474bd3a1b2bd-r-9f157451] time="2023-1
1-06T13:00:50Z" level=info msg="New connection from: 10.42.3.235:59868" func="rpc.(*DataServer).listenAndServeTCP" file="dataserver.go:57
"
containers/instance-manager-554c586ceba126a4394b9acba2421017_longhorn-system_instance-manager-d21cda831661a540310183bab441cdd0b7fa8504cbc
3e1bc1341186afd9ea2d9.log:2023-11-06T14:00:51.674015967+01:00 stderr F [pvc-53193e0a-d309-42f1-96f7-8e8be736bbef-r-29d25b3b] time="2023-1
1-06T13:00:51Z" level=info msg="New connection from: 10.42.3.235:38580" func="rpc.(*DataServer).listenAndServeTCP" file="dataserver.go:57
"
containers/instance-manager-554c586ceba126a4394b9acba2421017_longhorn-system_instance-manager-d21cda831661a540310183bab441cdd0b7fa8504cbc
3e1bc1341186afd9ea2d9.log:2023-11-06T14:00:52.086076435+01:00 stderr F [pvc-f1ac784b-0c9c-44ab-b86a-0a858694b6d9-r-906e43c3] time="2023-1
1-06T13:00:52Z" level=info msg="New connection from: 10.42.3.235:33000" func="rpc.(*DataServer).listenAndServeTCP" file="dataserver.go:57
"
containers/instance-manager-554c586ceba126a4394b9acba2421017_longhorn-system_instance-manager-d21cda831661a540310183bab441cdd0b7fa8504cbc
3e1bc1341186afd9ea2d9.log:2023-11-06T14:00:53.723757511+01:00 stderr F [pvc-4bd54b2f-0040-44b6-b6f8-b36459fe3d4d-r-0f18a6e8] time="2023-1
1-06T13:00:53Z" level=info msg="New connection from: 10.42.3.235:34508" func="rpc.(*DataServer).listenAndServeTCP" file="dataserver.go:57
"
```

## To Reproduce

I don't know, I found this in logs, but I'm not sure if there is anything specific on my system..

## Expected behavior

Probably longhorn should refresh list of those endpoints and remove those invalid. 

## Support bundle for troubleshooting

If that is needed I can send that to support e-mail. 

## Environment

 - Longhorn version: 1.5.2
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): helm
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: k3s 1.28.4
   - Number of management node in the cluster: 1
   - Number of worker node in the cluster: 4
 - Node config
   - OS type and version: debian 12 bookworm
   - Kernel version: latest from apt, 6.1.0-13-amd64
   - CPU per node: 4 
   - Memory per node: mostly 8GB, moving to 64GB
   - Disk type(e.g. SSD/NVMe/HDD): ssd+nvme
   - Network bandwidth between the nodes: 1Gbit
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): bare
 - Number of Longhorn volumes in the cluster: 50
 - Impacted Longhorn resources: ?
   - Volume names: ?

## Additional context

Same thing was on 1.27.x, upgrade changed nothing regarding this.

**Comment by PhanLe1010** [comment]
> Dec 19 13:19:36 k3s-node-2 iscsid[3160778]: connect to 10.42.3.235:3260 failed (No route to host)
> There is no such ip in k3s

Looks like the orphan tgt target leftover after the instance-manager pod is restart. It looks harmless to me. Any thoughts @derekbit @shuo-wu

**Comment by PhanLe1010** [comment]
> Probably longhorn should refresh list of those endpoints and remove those invalid.

Could you point me to how to refesh the list of iscsi sessions/tagets ?

**Comment by dominch** [comment]
> Could you point me to how to refesh the list of iscsi sessions/tagets ?

Invalid nodes/sessions are referenced on iscsi config, I could see such endpoint here
```
iscsiadm -m session
```
I have tried to delete that one manually with
```
iscsiadm -m node -o delete -T <iqn>
```
or delete target: 
```
iscsiadm -m discoverydb -t sendtargets -p <IP>:<port> -o delete
```
and node: 
```
 iscsiadm -m node -T <iqn> -p <ip address>:<port number> -u
```
Next step was to delete node config from: /etc/iscsi/nodes and restarting iscsd - but there are still same errors, 
I found out that they come from instance manager and probably there is yet another config that is managed and synced with node iscsid,

**Comment by PhanLe1010** [comment]
Hi @dominch Could you please send us [a support bundle](https://longhorn.io/docs/1.5.3/advanced-resources/support-bundle/) to `longhorn-support-bundle@suse.com`? I would like to double check if this is something we can enhance from Longhorn side

**Comment by PhanLe1010** [comment]
> I found out that they come from instance manager and probably there is yet another config that is managed and synced with node iscsid,

Sorry, could you give more details about this part?

**Comment by PhanLe1010** [comment]
From the code flow, I see this issue might potentially happen in theory when engine fail to logout and clean up the session (i.e.,when this function failed https://github.com/longhorn/longhorn-engine/blob/5ac6eec696db9b18ec90cac3d0aec090798ea23a/vendor/github.com/longhorn/go-iscsi-helper/iscsidev/iscsi.go#L258-L328)  during engine shutdown. However, I am not able to trigger it yet

**Comment by PhanLe1010** [comment]
Update:
I am able to see a similar (but not quite the same error logs) by:
1. Create a volume and attach it to node-1
1. exec into the instance-manager on node-1. List the PID of the engine process by `ps -aux`. For example, in my case it is `19`:
    ```
    instance-manager-8ec236b287c41733df5f2720fc624026:/ # ps -aux
    USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
    root           1  0.0  0.0   2520   656 ?        Ss   23:52   0:00 /tini -- instance-manager --debug daemon --listen 0.0.0.0:8500
    root           7  0.6  0.6 1468960 25816 ?       Sl   23:52   0:01 longhorn-instance-manager --debug daemon --listen 0.0.0.0:8500
    root          10  0.0  0.0 1134168 3172 ?        Sl   23:52   0:00 tgtd -f
    root          11  0.0  0.0   2536   700 ?        S    23:52   0:00 tee /var/log/tgtd.log
    root          19  0.3  0.5 1248384 23392 ?       Sl   23:52   0:00 /engine-binaries/longhornio-longhorn-engine-v1.5.3/longhorn --engine-instance-name testvol-e-0 controller 
    root         106  0.0  0.6 1248128 25176 ?       Sl   23:52   0:00 /host/var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.5.3/longhorn --volume-name testvol re
    root         111  0.0  0.5 1100664 23728 ?       Sl   23:52   0:00 /host/var/lib/longhorn/engine-binaries/longhornio-longhorn-engine-v1.5.3/longhorn --volume-name testvol sy
    root         130  1.2  0.1   8988  4288 pts/4    Ss   23:54   0:00 bash
    root         158 83.3  0.0  15780  3956 pts/4    R+   23:54   0:00 ps -aux
    ```
1. Hard kill the engine process without leaving it a chance to logout and cleanup the iscsid sessions `kill -9 19`
1. At the same time, detach the volume by `k edit -n longhorn-system volumeattachments.longhorn.io testvol` ad clear the spec.attachmenttickets. NOTE: this step have to be quick, otherwise the volume will reattach to the node and will cleanup the orphan iscsid session 
1. Delete the instance-manager pod
1. See the iscsid error log:
```
Dec 28 23:55:51 phan-v607-engine-image-pool2-e1b8ff09-ktjfk iscsid: cannot make a connection to 10.42.18.229:3260 (-1,22)
Dec 28 23:55:54 phan-v607-engine-image-pool2-e1b8ff09-ktjfk iscsid: cannot make a connection to 10.42.18.229:3260 (-1,22)
Dec 28 23:55:57 phan-v607-engine-image-pool2-e1b8ff09-ktjfk iscsid: cannot make a connection to 10.42.18.229:3260 (-1,22)
Dec 28 23:56:00 phan-v607-engine-image-pool2-e1b8ff09-ktjfk iscsid: cannot make a connection to 10.42.18.229:3260 (-1,22)
Dec 28 23:56:03 phan-v607-engine-image-pool2-e1b8ff09-ktjfk iscsid: cannot make a connection to 10.42.18.229:3260 (-1,22)
Dec 28 23:56:06 phan-v607-engine-image-pool2-e1b8ff09-ktjfk iscsid: cannot make a connection to 10.42.18.229:3260 (-1,22)
Dec 28 23:56:09 phan-v607-engine-image-pool2-e1b8ff09-ktjfk iscsid: cannot make a connection to 10.42.18.229:3260 (-1,22)
...
```
1. The Longhorn block device is still hanging around on this node
```
root@phan-v607-engine-image-pool2-e1b8ff09-ktjfk:~# ls /dev/longhorn/
testvol
```

**Comment by PhanLe1010** [comment]
TODO: Longhorn team need to document a workaround to get rid of the annoying logs

**Comment by dominch** [comment]
This is still an issue in longhorn v1.6 
any workaround to clear incorrect entries?

**Comment by Freundschaft** [comment]
I have a similar problem where it seems that the volumes that have the wrong ip set to them are not mounting correctly, is there some way to manually set the ip address to the right target?

**Comment by serhiynovos** [comment]
Similar for me longhorn 1.6.2
<img width="1014" alt="Знімок екрана 2024-06-27 о 16 50 06" src="https://github.com/longhorn/longhorn/assets/9201885/bf3f70ff-bd85-42ad-a1d9-5b535b75f171">

**Comment by derekbit** [comment]
cc @PhanLe1010

**Comment by serhiynovos** [comment]
> Similar for me longhorn 1.6.2
> 
> <img width="1014" alt="Знімок екрана 2024-06-27 о 16 50 06" src="https://github.com/longhorn/longhorn/assets/9201885/bf3f70ff-bd85-42ad-a1d9-5b535b75f171">
> 
> 

In my case I see this error only on nodes which have schedulable disks. I noticed this error when I was testing how my cluster tolerate to server/etcd nodes down and up. After i stopped 2 of 3 server nodes to make cluster unstable and then I started one by one nodes back I started seeing this error on worker nodes

**Comment by PhanLe1010** [comment]
@serhiynovos @Freundschaft @dominch Can you try this workaround:

1. Identidy the IP from the log. For example, `10.42.82.27` from this log:
```
Jun 28 01:25:41 phan-v672-pool2-1967f397-tprqc iscsid[17303]: cannot make a connection to 10.42.82.27:3260 (-1,22)
```
1. List all active sessions and find the one with the same IP as showing in the logs. For example:
```
root@phan-v672-pool2-1967f397-tprqc:~# sudo iscsiadm -m session -o show
tcp: [5] 10.42.82.27:3260,1 iqn.2019-10.io.longhorn:testvol (non-flash)
```
1. Log out of these sessions:
```
# Replace the targetname and IP with your actual values

root@phan-v672-pool2-1967f397-tprqc:~# iscsiadm --mode node --targetname iqn.2019-10.io.longhorn:testvol --portal 10.42.82.27:3260 --logout 
Logging out of session [sid: 5, target: iqn.2019-10.io.longhorn:testvol, portal: 10.42.82.27,3260]
Logout of [sid: 5, target: iqn.2019-10.io.longhorn:testvol, portal: 10.42.82.27,3260] successful.
```
1. Check if the log disappear

**Comment by PhanLe1010** [comment]
I think the goal of this ticket should be to document a workaround to get rid of the annoying logs.

To automatically cleanup, I think we would have to do it in the ticket https://github.com/longhorn/longhorn/issues/6764 . In that ticket, when we cleanup the orphan process, we can also cleaup iscsi session and tgt target. 

WDYT @ejweber @james-munson @derekbit

**Comment by serhiynovos** [comment]
@PhanLe1010 Thanks, I'll try steps provided by you.

**Comment by dominch** [comment]
@PhanLe1010 Thanks for Your time and commands, 
right now I don't see anything in that state on my cluster, I'll certainly come back here to try Your suggestion when I'll notice it again,

**Comment by ejweber** [comment]
> I think the goal of this ticket should be to document a KB article workaround to get rid of the annoying logs.
> 
> To automatically cleanup, I think we would have to do it in the ticket #6764 . In that ticket, when we cleanup the orphan process, we can also cleaup iscsi session and tgt target.

I agree. If the steps provided work for @serhiynovos (and @dominch if their environment requires it sometime soon), we can document them and follow up in #6764.

**Comment by derekbit** [comment]
> I think the goal of this ticket should be to document a KB article workaround to get rid of the annoying logs.
> 
> To automatically cleanup, I think we would have to do it in the ticket #6764 . In that ticket, when we cleanup the orphan process, we can also cleaup iscsi session and tgt target.
> 
> WDYT @ejweber @james-munson @derekbit

Agree as well.

**Comment by PhanLe1010** [comment]
Any update @dominch and @serhiynovos ?

**Comment by PhanLe1010** [comment]
## Test plan:

No specific test plan is needed for this ticket. Just make sure @dominch and @serhiynovos are happy with the workaround

**Comment by serhiynovos** [comment]
@PhanLe1010 so far I don't see these logs

**Comment by roger-ryao** [comment]
**Verified on master-head 20240704**

- longhorn `master-head` https://github.com/longhorn/longhorn/commit/c3e1f66bd67f5eeec837a5ee6522ac17391031ff

**The test steps**
https://longhorn.io/kb/troubleshooting-orphan-iscsi-session-error/

**Result**
1. After following up on https://github.com/longhorn/longhorn/issues/7386#issuecomment-1871633880, I could reproduce the issue on my side.
2. After following up on https://github.com/longhorn/longhorn/issues/7386#issuecomment-2195976658, the error log disappears.

```console
root@ryao-test-w1-81bb3f48-vrxvz:/home/ubuntu# journalctl -u iscsid -f
Jul 04 08:08:02 ryao-test-w1-81bb3f48-vrxvz iscsid[76521]: connect to 10.42.2.17:3260 failed (No route to host)
Jul 04 08:08:09 ryao-test-w1-81bb3f48-vrxvz iscsid[76521]: connect to 10.42.2.17:3260 failed (No route to host)
Jul 04 08:08:16 ryao-test-w1-81bb3f48-vrxvz iscsid[76521]: connect to 10.42.2.17:3260 failed (No route to host)
Jul 04 08:08:23 ryao-test-w1-81bb3f48-vrxvz iscsid[76521]: connect to 10.42.2.17:3260 failed (No route to host)
Jul 04 08:08:30 ryao-test-w1-81bb3f48-vrxvz iscsid[76521]: connect to 10.42.2.17:3260 failed (No route to host)
Jul 04 08:08:37 ryao-test-w1-81bb3f48-vrxvz iscsid[76521]: connect to 10.42.2.17:3260 failed (No route to host)
Jul 04 08:08:44 ryao-test-w1-81bb3f48-vrxvz iscsid[76521]: connect to 10.42.2.17:3260 failed (No route to host)
Jul 04 08:08:51 ryao-test-w1-81bb3f48-vrxvz iscsid[76521]: connect to 10.42.2.17:3260 failed (No route to host)
Jul 04 08:08:58 ryao-test-w1-81bb3f48-vrxvz iscsid[76521]: connect to 10.42.2.17:3260 failed (No route to host)
Jul 04 08:09:01 ryao-test-w1-81bb3f48-vrxvz iscsid[76521]: Connection4:0 to [target: iqn.2019-10.io.longhorn:vol-0, portal: 10.42.2.17,3260] through [iface: default] is shutdown.

root@ryao-test-w1-81bb3f48-vrxvz:/home/ubuntu# date
Thu Jul  4 08:19:31 UTC 2024
```

**Comment by dominch** [comment]
I just verified this, 

Yesterday I tried to reproduce iscsi problem on one of nodes, upgraded k3s, rebooted it, used k3s-killall script - no luck, error log was clear. @serhiynovos gave me the clue - I needed to stop etcd node to get this problem, as described - with k3s-killall problem reappeared on three nodes. 

Checked manually solution on first node, I got one failed IP in logs and about 9 sessions with this particular IP. This took me while to clear everything connected with problematic IP (I assumed that I need to logout all sessions by this IP), workaround works for sure with same outup as in https://github.com/longhorn/longhorn/issues/7386#issuecomment-2208394278

On second node I decided to parse logs for ips and then iscsiadm for iqn, here is log for problem: 
```
root@node3:~# journalctl --since="20 minutes ago" --no-pager -u iscsid | tail
Jul 04 14:20:21 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
Jul 04 14:20:28 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
Jul 04 14:20:35 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
Jul 04 14:20:42 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
Jul 04 14:20:49 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
Jul 04 14:20:56 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
Jul 04 14:21:03 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
Jul 04 14:21:10 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
Jul 04 14:21:17 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
Jul 04 14:21:24 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
```

then, parsing problematic ips: 

```
root@node3:~# IFS=$'\n'; ips=`journalctl --since="20 minutes ago" --no-pager -u iscsid | grep "No route to host" | grep -o '[0-9]\{1,3\}\.[0-9]\{1,3\}\.[0-9]\{1,3\}\.[0-9]\{1,3\}' | uniq`
root@node3:~# echo $ips
10.42.2.16
```

then loop for each ip and sessions to logout: 

```
for ip in "${ips[@]}"; do echo "host: $ip"; IFS=$'\n' iqn=(`iscsiadm -m node show | grep "$ip" | awk '{print $2}'`); echo "endpoints: ${#iqn[@]}"; for iq in "${iqn[@]}"; do iscsiadm -m node -T $iq -p $ip --logout ; done; done
```

result: 

```
host: 10.42.2.16
endpoints: 1
Logging out of session [sid: 10, target: iqn.2019-10.io.longhorn:pvc-a476ac01-2d14-4295-a033-1c38d48f76b4, portal: 10.42.2.16,3260]
Logout of [sid: 10, target: iqn.2019-10.io.longhorn:pvc-a476ac01-2d14-4295-a033-1c38d48f76b4, portal: 10.42.2.16,3260] successful.
```

and journal for this action: 

```
Jul 04 14:22:55 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
Jul 04 14:23:02 node3 iscsid[635]: connect to 10.42.2.16:3260 failed (No route to host)
Jul 04 14:23:05 node3 iscsid[635]: Connection10:0 to [target: iqn.2019-10.io.longhorn:pvc-a476ac01-2d14-4295-a033-1c38d48f76b4, portal: 10.42.2.16,3260] through [iface: default] is shutdown.
```

The only thing that is missing at this point it's node still on list of scsiadm: 

```
root@node3:~# iscsiadm -m node show
10.42.2.15:3260,1 iqn.2019-10.io.longhorn:pvc-070cbea9-0412-4bec-9de7-8d50dbaa497f
10.42.2.15:3260,1 iqn.2019-10.io.longhorn:pvc-09a52ecd-1fe3-4b38-bb12-d3bff928428c
10.42.2.15:3260,1 iqn.2019-10.io.longhorn:pvc-120df2d5-26a4-452c-8601-a24ba29f8381
10.42.2.15:3260,1 iqn.2019-10.io.longhorn:pvc-34ade422-fec9-4bab-98bf-4555e1822aff
10.42.2.15:3260,1 iqn.2019-10.io.longhorn:pvc-4b228468-98eb-4b56-9c11-9c93b16d1388
10.42.2.15:3260,1 iqn.2019-10.io.longhorn:pvc-53193e0a-d309-42f1-96f7-8e8be736bbef
10.42.2.15:3260,1 iqn.2019-10.io.longhorn:pvc-5af5698b-414d-4478-be04-04028820fe88
10.42.2.26:3260,1 iqn.2019-10.io.longhorn:pvc-5d7cafff-027d-4e23-9775-b7725f0dca38
10.42.2.183:3260,1 iqn.2019-10.io.longhorn:pvc-77f6e89e-815b-403e-aea6-21a18be9ef2b
10.42.2.15:3260,1 iqn.2019-10.io.longhorn:pvc-9644e42f-bc49-469a-a29e-4559fc4bdcec
10.42.2.15:3260,1 iqn.2019-10.io.longhorn:pvc-97edec23-6d91-47be-b791-d4a83adbfd0a
10.42.2.16:3260,1 iqn.2019-10.io.longhorn:pvc-a476ac01-2d14-4295-a033-1c38d48f76b4 (*** this one ***)
10.42.2.15:3260,1 iqn.2019-10.io.longhorn:pvc-c371f7d6-8855-477f-badf-252167271ffb
10.42.2.15:3260,1 iqn.2019-10.io.longhorn:pvc-ffa49840-3bee-4981-ae00-be59ea1b4a21
```

For this I needed: 

```
iscsiadm -m node -T iqn.2019-10.io.longhorn:pvc-a476ac01-2d14-4295-a033-1c38d48f76b4 -o delete
```

Full script to find broken ips in journalctl, then nodes for that ip in iscsiadm, then logout each session and remove node: 

```
IFS=$'\n'; ips=`journalctl --since="20 minutes ago" --no-pager -u iscsid | grep "No route to host" | grep -o '[0-9]\{1,3\}\.[0-9]\{1,3\}\.[0-9]\{1,3\}\.[0-9]\{1,3\}' | uniq`
for ip in "${ips[@]}"; do echo "host: $ip"; IFS=$'\n' iqn=(`iscsiadm -m node show | grep "$ip" | awk '{print $2}'`); echo "endpoints: ${#iqn[@]}"; for iq in "${iqn[@]}"; do iscsiadm -m node -T $iq -p $ip --logout ; iscsiadm -m node -T $iq -o delete; done; done
```

**Comment by axc274** [comment]
Wanted to add context from our environment that may help scope the fix discussion.

We've been hitting this consistently on k3s clusters running **Longhorn 1.9.2**. The root trigger in our case isn't limited to node reboots — we see orphaned iscsid sessions after **any event that causes instance-manager to be recreated with a new pod IP**: scheduler evictions, disk pressure evictions, and regular rolling redeployments, in addition to node reboots. The common thread is simply that the instance-manager pod IP changes and iscsid is left holding a session to the old address.

The workaround documented by @dominch in [this comment](https://github.com/longhorn/longhorn/issues/7386#issuecomment-2208394278) (surgical `iscsiadm --logout` + delete rather than `systemctl stop iscsid`) is solid and worth promoting to the KB article as the preferred approach, since `systemctl stop iscsid` kills active sessions too.

On the fix side: @PhanLe1010 proposed extending #6764 to include iSCSI session cleanup alongside engine/replica instance cleanup. That issue closed as part of v1.9.0, but the released `orphan-resource-auto-deletion` setting appears to cover `engineInstance` and `replicaInstance` only — it's not clear whether the iSCSI session cleanup proposed in [this comment](https://github.com/longhorn/longhorn/issues/6764#issuecomment-2195984587) was actually implemented. We are running 1.9.2 and still seeing the issue, which suggests it was not.

Can someone confirm whether v1.9.0+ automatically cleans up orphaned iSCSI sessions to stale instance-manager IPs, or whether that scope was deferred?

If it was deferred, this is the ask: add iSCSI session cleanup to `orphan-resource-auto-deletion` so that when instance-manager restarts and gets a new IP, any stale sessions to the old IP are automatically logged out and deleted without manual intervention. The logic @PhanLe1010 described — longhorn-manager periodically sending the current IM IP list to IM, which then cleans up sessions to non-existent IPs — seems like the right approach.

**Comment by shuo-wu** [comment]
> On the fix side: @PhanLe1010 proposed extending https://github.com/longhorn/longhorn/issues/6764 to include iSCSI session cleanup alongside engine/replica instance cleanup. That issue closed as part of v1.9.0, but the released orphan-resource-auto-deletion setting appears to cover engineInstance and replicaInstance only — it's not clear whether the iSCSI session cleanup proposed in https://github.com/longhorn/longhorn/issues/6764#issuecomment-2195984587 was actually implemented. We are running 1.9.2 and still seeing the issue, which suggests it was not.
>
> Can someone confirm whether v1.9.0+ automatically cleans up orphaned iSCSI sessions to stale instance-manager IPs, or whether that scope was deferred?

I don't think [the auto engine/replica orphan instance cleanup](https://github.com/longhorn/longhorn-manager/blob/5ddaa369fa7e3b0da9ee9fd42dbec24e25f793f6/controller/orphan_controller.go#L445) will handle these kinds of stale iSCSI connections, especially when there is no associated engine process for the connection. 
cc @COLDTURNIP

**Comment by DerRockWolf** [comment]
I'm also seeing this happening frequently now, also running on 1.9.2
It seems like it also prevents volumes from successfully getting mounted in some cases.
